hamilton/lifecycle/base.py

```diff
@@ -745,11 +745,20 @@ def __init__(self, *adapters: LifecycleAdapter):
 
         :param adapters: Adapters to group together
         """
-        self._adapters = list(adapters)
+        self._adapters = self._uniqify_adapters(adapters)
         self.sync_hooks, self.async_hooks = self._get_lifecycle_hooks()
         self.sync_methods, self.async_methods = self._get_lifecycle_methods()
         self.sync_validators = self._get_lifecycle_validators()
 
+    def _uniqify_adapters(self, adapters: List[LifecycleAdapter]) -> List[LifecycleAdapter]:
+        """Removes duplicate adapters from the list of adapters -- this often happens on how they're passed in
```

Comment:
how would this happen in practice?

Reply:
Yeah, it's not great, but deduplicating makes life easier. Will add a TODO.

What happens is that we think about adapters as keys (hook/method names), whereas they are objects. When we do some cleverness (separate them out, etc...) they end up getting merged back in. Might be worth digging in more but for now this is OK.